### PR TITLE
DT-2926: ✨ Replace nomis email lookup by call to nomis user api

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/model/NomisApiUserDetails.kt [hunk 1]
```diff
@@ -8,7 +8,7 @@ data class NomisApiUserDetails @JsonCreator constructor(
   @JsonProperty("username") val username: String,
   @JsonProperty("lastName") val surname: String,
   @JsonProperty("firstName") val firstName: String,
-  @JsonProperty("primaryEmail") val email: String,
+  @JsonProperty("primaryEmail") val email: String?,
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/model/NomisApiUserDetails.kt:11]: some users don't have emails in nomis

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/repository/StaffUserAccountRepository.kt [hunk 2]
```diff
@@ -7,14 +7,6 @@ import uk.gov.justice.digital.hmpps.oauth2server.nomis.model.NomisUserPersonDeta
 
 @Suppress("SqlResolve")
 interface StaffUserAccountRepository : CrudRepository<NomisUserPersonDetails, String> {
-  @Query(
-    value = """select distinct s.*, '' as SPARE4 from STAFF_USER_ACCOUNTS s
-    inner join INTERNET_ADDRESSES i on i.owner_id = s.staff_id and owner_class = 'STF'
-    where internet_address_class = 'EMAIL' and i.internet_address = ?""",
-    nativeQuery = true
-  )
-  fun findAllNomisUsersByEmailAddress(email: String): List<NomisUserPersonDetails>
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/repository/StaffUserAccountRepository.kt:16]: removed and call nomis user roles api instead

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt [hunk 3]
```diff
@@ -25,9 +27,18 @@ import uk.gov.justice.digital.hmpps.oauth2server.security.ReusedPasswordExceptio
...
   @Qualifier("nomisUserWebClient") private val nomisUserWebClient: WebClient,
+  @Value("\${nomis.enabled:false}") private val nomisEnabled: Boolean,
   private val objectMapper: ObjectMapper,
 ) {
+  private companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
   fun changePassword(username: String, password: String) {
+    if (!nomisEnabled) {
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt:38]: altering code to match delius api for consistency

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt [hunk 4]
```diff
@@ -45,32 +56,49 @@ class NomisUserApiService(
...
+    val userDetails = webClient.post().uri {
       it.path("/users/user")
         .queryParam("email", emailAddress)
         .build()
     }
+      .bodyValue(usernames)
       .retrieve()
-      .bodyToMono(NomisUserList::class.java)
-      .block()
+      .bodyToMono(object : ParameterizedTypeReference<List<NomisApiUserDetails>>() {})
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt:92]: simplifying code by not adding unnecessary extra class

**Review by andymarke — APPROVED** [review]
LGTM

**Review by Mjwillis — APPROVED** [review]
LGTM

**Pull request merged**